File: mobile/composeApp/src/commonMain/kotlin/io/j3mobile/protocol/Git.kt
package io.j3mobile.protocol

import kotlinx.serialization.Serializable

/** Input for a git status request. */
@Serializable
data class GitStatusInput(
    val cwd: String,
)

/** Result of a git status request. */
@Serializable
data class GitStatusResult(
    val branch: String? = null,
    val hasWorkingTreeChanges: Boolean,
    val workingTree: GitWorkingTree,
    val hasUpstream: Boolean,
    val aheadCount: Int,
    val behindCount: Int,
    val pr: GitPullRequest? = null,
)

/** Summarizes changes in the working tree. */
@Serializable
data class GitWorkingTree(
    val files: List<GitFileChange> = emptyList(),
    val insertions: Int,
    val deletions: Int,
)

/** Describes a single changed file. */
@Serializable
data class GitFileChange(
    val path: String,
    val insertions: Int,
    val deletions: Int,
)

/** Describes a pull request associated with a branch. */
@Serializable
data class GitPullRequest(
    val number: Int,
    val title: String,
    val url: String,
    val baseBranch: String,
    val headBranch: String,
    val state: String,
)

/** Input for a git pull request. */
@Serializable
data class GitPullInput(
    val cwd: String,
)

/** Result of a git pull operation. */
@Serializable
data class GitPullResult(
    val success: Boolean,
    val message: String,
)

/** Input for listing branches. */
@Serializable
data class GitListBranchesInput(
    val cwd: String,
)

/** Describes a git branch. */
@Serializable
data class GitBranch(
    val name: String,
    val isCurrent: Boolean,
    val isRemote: Boolean,
    val commit: String,
)

/** Result of listing branches. */
@Serializable
data class GitListBranchesResult(
    val branches: List<GitBranch> = emptyList(),
)

/** Input for creating a branch. */
@Serializable
data class GitCreateBranchInput(
    val cwd: String,
    val branchName: String,
    val startPoint: String? = null,
)

/** Input for checking out a branch. */
@Serializable
data class GitCheckoutInput(
    val cwd: String,
    val branchName: String,
)

/** Input for initializing a git repository. */
@Serializable
data class GitInitInput(
    val cwd: String,
)

/** Input for creating a worktree. */
@Serializable
data class GitCreateWorktreeInput(
    val cwd: String,
    val path: String,
    val branchName: String? = null,
)

/** Result of creating a worktree. */
@Serializable
data class GitCreateWorktreeResult(
    val path: String,
    val branch: String,
)

/** Input for removing a worktree. */
@Serializable
data class GitRemoveWorktreeInput(
    val cwd: String,
    val path: String,
)

/** Input for running a stacked git action. */
@Serializable
data class GitRunStackedActionInput(
    val actionId: String,
    val cwd: String,
    val action: String,
    val commitMessage: String? = null,
    val featureBranch: Boolean? = null,
    val filePaths: List<String>? = null,
    val modelSelection: ModelSelection,
)

/** Result of a stacked git action. */
@Serializable
data class GitRunStackedActionResult(
    val action: String,
    val branch: GitStackedActionBranch? = null,
    val commit: GitStackedActionCommit? = null,
    val push: GitStackedActionPush? = null,
    val pr: GitStackedActionPullRequest? = null,
)

/** Branch result of a stacked action. */
@Serializable
data class GitStackedActionBranch(
    val name: String,
    val created: Boolean,
)

/** Commit result of a stacked action. */
@Serializable
data class GitStackedActionCommit(
    val sha: String,
    val message: String,
)

/** Push result of a stacked action. */
@Serializable
data class GitStackedActionPush(
    val remote: String,
    val branch: String,
)

/** PR result of a stacked action. */
@Serializable
data class GitStackedActionPullRequest(
    val number: Int,
    val url: String,
    val title: String,
)

/** Push event for git action progress. */
@Serializable
data class GitActionProgressEvent(
    val actionId: String,
    val cwd: String,
    val action: String,
    val kind: String,
    val message: String? = null,
    val step: String? = null,
    val progress: Int? = null,
    val total: Int? = null,
    val error: String? = null,
)

File: server/internal/protocol/git.go
package protocol

// GitStatusInput is the input for a git status request.
type GitStatusInput struct {
	Cwd string `json:"cwd"`
}

// GitStatusResult is the result of a git status request.
type GitStatusResult struct {
	Branch                *string         `json:"branch,omitempty"`
	HasWorkingTreeChanges bool            `json:"hasWorkingTreeChanges"`
	WorkingTree           GitWorkingTree  `json:"workingTree"`
	HasUpstream           bool            `json:"hasUpstream"`
	AheadCount            int             `json:"aheadCount"`
	BehindCount           int             `json:"behindCount"`
	PR                    *GitPullRequest `json:"pr,omitempty"`
}

// GitWorkingTree summarizes changes in the working tree.
type GitWorkingTree struct {
	Files      []GitFileChange `json:"files"`
	Insertions int             `json:"insertions"`
	Deletions  int             `json:"deletions"`
}

// GitFileChange describes a single changed file.
type GitFileChange struct {
	Path       string `json:"path"`
	Insertions int    `json:"insertions"`
	Deletions  int    `json:"deletions"`
}

// GitPullRequest describes a pull request associated with a branch.
type GitPullRequest struct {
	Number     int    `json:"number"`
	Title      string `json:"title"`
	URL        string `json:"url"`
	BaseBranch string `json:"baseBranch"`
	HeadBranch string `json:"headBranch"`
	State      string `json:"state"`
}

// GitPullInput is the input for a git pull request.
type GitPullInput struct {
	Cwd string `json:"cwd"`
}

// GitPullResult is the result of a git pull operation.
type GitPullResult struct {
	Success bool   `json:"success"`
	Message string `json:"message"`
}

// GitListBranchesInput is the input for listing branches.
type GitListBranchesInput struct {
	Cwd string `json:"cwd"`
}

// GitBranch describes a git branch.
type GitBranch struct {
	Name      string `json:"name"`
	IsCurrent bool   `json:"isCurrent"`
	IsRemote  bool   `json:"isRemote"`
	Commit    string `json:"commit"`
}

// GitListBranchesResult is the result of listing branches.
type GitListBranchesResult struct {
	Branches []GitBranch `json:"branches"`
}

// GitCreateBranchInput is the input for creating a branch.
type GitCreateBranchInput struct {
	Cwd        string  `json:"cwd"`
	BranchName string  `json:"branchName"`
	StartPoint *string `json:"startPoint,omitempty"`
}

// GitCheckoutInput is the input for checking out a branch.
type GitCheckoutInput struct {
	Cwd        string `json:"cwd"`
	BranchName string `json:"branchName"`
}

// GitInitInput is the input for initializing a git repository.
type GitInitInput struct {
	Cwd string `json:"cwd"`
}

// GitCreateWorktreeInput is the input for creating a worktree.
type GitCreateWorktreeInput struct {
	Cwd        string  `json:"cwd"`
	Path       string  `json:"path"`
	BranchName *string `json:"branchName,omitempty"`
}

// GitCreateWorktreeResult is the result of creating a worktree.
type GitCreateWorktreeResult struct {
	Path   string `json:"path"`
	Branch string `json:"branch"`
}

// GitRemoveWorktreeInput is the input for removing a worktree.
type GitRemoveWorktreeInput struct {
	Cwd  string `json:"cwd"`
	Path string `json:"path"`
}

// GitRunStackedActionInput is the input for running a stacked git action.
type GitRunStackedActionInput struct {
	ActionId       string         `json:"actionId"`
	Cwd            string         `json:"cwd"`
	Action         string         `json:"action"`
	CommitMessage  *string        `json:"commitMessage,omitempty"`
	FeatureBranch  *bool          `json:"featureBranch,omitempty"`
	FilePaths      []string       `json:"filePaths,omitempty"`
	ModelSelection ModelSelection `json:"modelSelection"`
}

// GitRunStackedActionResult is the result of a stacked git action.
type GitRunStackedActionResult struct {
	Action string                       `json:"action"`
	Branch *GitStackedActionBranch      `json:"branch,omitempty"`
	Commit *GitStackedActionCommit      `json:"commit,omitempty"`
	Push   *GitStackedActionPush        `json:"push,omitempty"`
	PR     *GitStackedActionPullRequest `json:"pr,omitempty"`
}

// GitStackedActionBranch describes the branch result of a stacked action.
type GitStackedActionBranch struct {
	Name    string `json:"name"`
	Created bool   `json:"created"`
}

// GitStackedActionCommit describes the commit result of a stacked action.
type GitStackedActionCommit struct {
	SHA     string `json:"sha"`
	Message string `json:"message"`
}

// GitStackedActionPush describes the push result of a stacked action.
type GitStackedActionPush struct {
	Remote string `json:"remote"`
	Branch string `json:"branch"`
}

// GitStackedActionPullRequest describes the PR result of a stacked action.
type GitStackedActionPullRequest struct {
	Number int    `json:"number"`
	URL    string `json:"url"`
	Title  string `json:"title"`
}

// GitActionProgressEvent is a push event for git action progress.
type GitActionProgressEvent struct {
	ActionId string  `json:"actionId"`
	Cwd      string  `json:"cwd"`
	Action   string  `json:"action"`
	Kind     string  `json:"kind"`
	Message  *string `json:"message,omitempty"`
	Step     *string `json:"step,omitempty"`
	Progress *int    `json:"progress,omitempty"`
	Total    *int    `json:"total,omitempty"`
	Error    *string `json:"error,omitempty"`
}
